resource-managers/kubernetes/core/src/test/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleDataIOSuite.scala

```diff
@@ -0,0 +1,219 @@
+
+      // Decommission all executors
+      val sched = sc.schedulerBackend.asInstanceOf[StandaloneSchedulerBackend]
+      sc.getExecutorIds().foreach { id =>
+        sched.killExecutor(id)
+      }
+      TestUtils.waitUntilExecutorsUp(sc, 1, 60000)
+      // Shuffle status are removed
+      eventually(timeout(60.second), interval(1.seconds)) {
+        assert(master.shuffleStatuses.keys.toSet == Set(0))
+        assert(master.shuffleStatuses(0).mapStatuses.forall(_ == null))
+      }
+
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(9, 10, 11))
+    }
+  }
+
+  test("Partial recompute shuffle data") {
+    sc = new SparkContext(conf)
+    withSpark(sc) { sc =>
+      val master = sc.env.mapOutputTracker.asInstanceOf[MapOutputTrackerMaster]
+      assert(master.shuffleStatuses.isEmpty)
+
+      val rdd = sc.parallelize(Seq((1, "one"), (2, "two"), (3, "three")), 3).groupByKey()
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0))
+      val loc1 = master.shuffleStatuses(0).mapStatuses(0).location
+      assert(master.shuffleStatuses(0).mapStatuses.forall(_.location == loc1))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(0, 1, 2))
+
+      // Reuse the existing shuffle data
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0))
+      assert(master.shuffleStatuses(0).mapStatuses.forall(_.location == loc1))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(0, 1, 2))
+
+      val rdd2 = sc.parallelize(Seq((4, "four"), (5, "five"), (6, "six"), (7, "seven")), 4)
+        .groupByKey()
+      rdd2.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0, 1))
+      assert(master.shuffleStatuses(0).mapStatuses.forall(_.location == loc1))
+      assert(master.shuffleStatuses(1).mapStatuses.forall(_.location == loc1))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(0, 1, 2))
+      assert(master.shuffleStatuses(1).mapStatuses.map(_.mapId).toSet == Set(9, 10, 11, 12))
+
+      // Decommission all executors
+      val sched = sc.schedulerBackend.asInstanceOf[StandaloneSchedulerBackend]
+      sc.getExecutorIds().foreach { id =>
+        sched.killExecutor(id)
+      }
+      TestUtils.waitUntilExecutorsUp(sc, 1, 60000)
+      // Shuffle status are removed
+      eventually(timeout(60.second), interval(1.seconds)) {
+        assert(master.shuffleStatuses.keys.toSet == Set(0, 1))
+        assert(master.shuffleStatuses(0).mapStatuses.forall(_ == null))
+        assert(master.shuffleStatuses(1).mapStatuses.forall(_ == null))
+      }
+
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0, 1))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(17, 18, 19))
+      assert(master.shuffleStatuses(1).mapStatuses.map(_.mapId).toSet == Set(9, 10, 11, 12))
```

Comment:
Why rdd is recomputed, but rdd2 is to recover from old shuffle status?

Reply:
This is a continuation of the previous test, `test("recompute is not blocked by the recovery")`.
We have two independent RDD `rdd` and `rdd2` and add is invoked first before recovery. `rdd2` is not used yet, but it's recovered.